python/ray/tune/log_sync.py

```diff
@@ -38,23 +42,64 @@ def wait_for_log_sync():
         syncer.wait()
 
 
-class _S3LogSyncer(object):
-    def __init__(self, local_dir, remote_dir):
+class _LogSyncer(object):
+    """Log syncer for tune.
+
+    This syncs files from workers to the local node, and optionally also from
+    the local node to a remote directory (e.g. S3)."""
+
+    def __init__(self, local_dir, remote_dir=None):
         self.local_dir = local_dir
         self.remote_dir = remote_dir
         self.last_sync_time = 0
         self.sync_process = None
-        print("Created S3LogSyncer for {} -> {}".format(local_dir, remote_dir))
+        self.local_ip = ray.services.get_node_ip_address()
+        self.worker_ip = None
+        print("Created LogSyncer for {} -> {}".format(local_dir, remote_dir))
+
+    def set_worker_ip(self, worker_ip):
+        """Set the worker ip to sync logs from."""
+
+        self.worker_ip = worker_ip
 
     def sync_if_needed(self):
         if time.time() - self.last_sync_time > 300:
             self.sync_now()
 
     def sync_now(self, force=False):
-        print(
-            "Syncing files from {} -> {}".format(
-                self.local_dir, self.remote_dir))
         self.last_sync_time = time.time()
+        if not self.worker_ip:
+            print(
+                "Worker ip unknown, skipping log sync for {}".format(
+                    self.local_dir))
+            return
+
+        if self.worker_ip == self.local_ip:
+            worker_to_local_sync_cmd = None  # don't need to rsync
+        else:
+            ssh_key = get_ssh_key()
+            ssh_user = get_ssh_user()
+            if ssh_key is None or ssh_user is None:
+                print(
+                    "Error: log sync requires cluster to be setup with "
+                    "`ray create_or_update`.")
+                return
+            if not distutils.spawn.find_executable("rsync"):
+                print("Error: log sync requires rsync to be installed.")
+                return
+            worker_to_local_sync_cmd = (
+                ("""rsync -avz -e "ssh -i '{}' -o ConnectTimeout=120s """
```

Comment:
this also doesn't work with Docker btw

Reply:
Yeah, it would be nice to have a more generalized way of running remote commands. Probably when kubernetes support lands we can expose a common run_cmd interface from the autoscaler.